Lib/dataclasses.py

```diff
@@ -1014,6 +1014,18 @@ def _process_class(cls, init, repr, eq, order, unsafe_hash, frozen,
     if init:
         # Does this class have a post-init function?
         has_post_init = hasattr(cls, _POST_INIT_NAME)
+        init_globals = globals
+        if has_dataclass_bases:
+            # If dataclass has other dataclass as a base type,
+            # it might have existing `__init__` method.
+            # We need to change its `globals`, because otherwise
+            # we might end up with unsolvable annotations. For example:
+            # `def __init__(self, d: collections.OrderedDict) -> None:`
+            # won't be able to resolve `collections.OrderedDict`
+            # with wrong `globals`, when placed in different modules. #45524
+            super_init = getattr(cls, '__init__', None)
+            if super_init is not None:
+                init_globals = getattr(super_init, '__globals__', globals)
```

Comment:
I am a little queasy about tinkering with `globals()` directly like this (though I can't point to any tests failing at all!). 

Perhaps we could make sure that `init_globals` is always a different dictionary to `globals()`, rather than being sometimes-a-different-dict-and-sometimes-actually-globals-itself? The following seems to pass tests just as well:

```suggestion
        init_globals = dict(globals)
        if has_dataclass_bases:
            # If dataclass has other dataclass as a base type,
            # it might have existing `__init__` method.
            # We need to change its `globals`, because otherwise
            # we might end up with unsolvable annotations. For example:
            # `def __init__(self, d: collections.OrderedDict) -> None:`
            # won't be able to resolve `collections.OrderedDict`
            # with wrong `globals`, when placed in different modules. #45524
            super_init = getattr(cls, '__init__', None)
            if super_init is not None:
                init_globals = getattr(super_init, '__globals__', init_globals )
```